Source/Core/Core/IOS/USB/Emulated/Skateboard.cpp

```diff
@@ -0,0 +1,348 @@
+{
+  return {s_interface_descriptor};
+}
+
+std::vector<EndpointDescriptor> SkateboardUSB::GetEndpoints(u8 config, u8, u8 alt) const
+{
+  return s_endpoint_descriptors;
+}
+
+bool SkateboardUSB::Attach()
+{
+  // This function gets called over and over,
+  // so don't print anything here.
+  return true;
+}
+
+bool SkateboardUSB::AttachAndChangeInterface(u8 interface)
+{
+  ERROR_LOG_FMT(IOS_USB, "FIXME: Skateboard AttachAndChangeInterface");
+  return true;
+}
+
+int SkateboardUSB::CancelTransfer(u8 endpoint)
+{
+  ERROR_LOG_FMT(IOS_USB, "FIXME: Skateboard CancelTransfer");
+  return 0;
+}
+
+int SkateboardUSB::ChangeInterface(const u8 interface)
+{
+  ERROR_LOG_FMT(IOS_USB, "FIXME: Skateboard ChangeInterface");
+  return 0;
+}
+
+int SkateboardUSB::GetNumberOfAltSettings(u8 interface)
+{
+  ERROR_LOG_FMT(IOS_USB, "FIXME: Skateboard GetNumberOfAltSettings");
+  return 0;
+}
+
+int SkateboardUSB::SetAltSetting(u8 alt_setting)
+{
+  ERROR_LOG_FMT(IOS_USB, "FIXME: Skateboard SetAltSetting");
+  return 0;
+}
+
+int SkateboardUSB::SubmitTransfer(std::unique_ptr<CtrlMessage> message)
+{
+  static constexpr u8 GET_DESCRIPTOR = 0x06;
+  static constexpr u8 HID_SET_REPORT = 0x09;
+  static constexpr u8 HID_SET_PROTOCOL = 0x0B;
+
+  if (message->request_type == 0x21 && message->request == HID_SET_REPORT)
+  {
+    // The game sends these once every 1.25 seconds.
+  }
+  else if (message->request_type == 0x80 && message->request == GET_DESCRIPTOR &&
+           (message->value == 0x301 || message->value == 0x302))
+  {
+    // TODO: string descriptors
+    // 1: "Licensed by Nintendo of America"
+    // 2: "Skateboard Controller"
+  }
+  else if (message->request_type == 0x21 && message->request == HID_SET_PROTOCOL)
+  {
+    // TODO: switch between boot and report protocol (not sure what that means)
+  }
+  else
+  {
+    auto buf = message->MakeBuffer(message->length);
+    ERROR_LOG_FMT(IOS_USB,
+                  "FIXME: CtrlMessage bRequestType={:02x} bRequest={:02x} wValue={:04x} "
+                  "wIndex={:04x} wLength={:04x}",
+                  message->request_type, message->request, message->value, message->index,
+                  message->length);
+  }
+  message->ScheduleTransferCompletion(message->length, 0);
+  return 0;
+}
+
+int SkateboardUSB::SubmitTransfer(std::unique_ptr<IntrMessage> message)
+{
+  if (message->endpoint != 0x81)
+    ERROR_LOG_FMT(IOS_USB, "Skateboard: unexpected IntrMessage");
+
+  const SkateboardHidReport report = m_inputs.GetState();
+  message->FillBuffer(reinterpret_cast<const u8*>(&report), sizeof(report));
+  message->ScheduleTransferCompletion(sizeof(report), 1000);
+  return 0;
+}
+
+int SkateboardUSB::SubmitTransfer(std::unique_ptr<BulkMessage> message)
+{
+  ERROR_LOG_FMT(IOS_USB, "Skateboard: unexpected BulkMessage");
+  return 0;
+}
+
+int SkateboardUSB::SubmitTransfer(std::unique_ptr<IsoMessage> message)
+{
+  ERROR_LOG_FMT(IOS_USB, "Skateboard: unexpected IsoMessage");
```

Comment:
I wonder if it makes sense to have default implementations of SubmitTransfer that do basically what you're doing here (log an error message and return success) - seems like most emulated devices won't need isochronous transfers for instance.